transport/httptransport/http_transport.go

```diff
@@ -121,41 +158,79 @@ func (t *transfer) emitEvent(ctx context.Context, evt types.TransportEvent, id u
 }
 
 func (t *transfer) execute(ctx context.Context) error {
-	// construct request
-	req, err := http.NewRequest("GET", t.tInfo.URL, nil)
-	if err != nil {
-		return fmt.Errorf("failed to create http req: %w", err)
+	b := &backoff.Backoff{
+		Min:    t.h.minBackOffWait,
+		Max:    t.h.maxBackoffWait,
+		Factor: t.h.backOffFactor,
+		Jitter: true,
 	}
 
-	// add range req to start reading from the last byte we have in the output file
-	req.Header.Set("Range", fmt.Sprintf("bytes=%d-", t.nBytesReceived))
-	// init the request with the transfer context
-	req = req.WithContext(ctx)
-	// open output file in append-only mode for writing
-	of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
-	if err != nil {
-		return fmt.Errorf("failed to open output file: %w", err)
-	}
-	defer of.Close()
+	for {
+		// construct request
+		req, err := http.NewRequest("GET", t.tInfo.URL, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create http req: %w", err)
+		}
+
+		st, err := os.Stat(t.dealInfo.OutputFile)
+		if err != nil {
+			return fmt.Errorf("failed to stat output file: %w", err)
+		}
+		t.nBytesReceived = st.Size()
+
+		// add range req to start reading from the last byte we have in the output file
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-", t.nBytesReceived))
+		// init the request with the transfer context
+		req = req.WithContext(ctx)
+		// open output file in append-only mode for writing
+		of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open output file: %w", err)
+		}
+		defer of.Close()
 
-	// start the http transfer
-	remaining := t.dealInfo.DealSize - t.nBytesReceived
-	if err := t.doHttp(ctx, req, of, remaining); err != nil {
-		// do not resume transfer if context has been cancelled
-		if ctx.Err() == context.Canceled || ctx.Err() == context.DeadlineExceeded {
-			return fmt.Errorf("transfer context err: %w", ctx.Err())
+		// start the http transfer
+		remaining := t.dealInfo.DealSize - t.nBytesReceived
+		if err = t.doHttp(ctx, req, of, remaining); err == nil {
+			// if there's no error, transfer was successful
+			break
+		}
+		_ = of.Close()
+		// do not resume transfer if context has been cancelled or if the context deadline has exceeded
+		if xerrors.Is(err, context.Canceled) || xerrors.Is(err, context.DeadlineExceeded) {
+			return fmt.Errorf("transfer context err: %w", err)
 		}
 
-		//
-		// TODO: resumption
-		return fmt.Errorf("failed to execute http transfer: %w", err)
+		// backoff-retry transfer if max number of attempts haven't been exhausted
+		nAttempts := b.Attempt() + 1
+		if nAttempts >= t.h.maxReconnectAttempts {
+			return fmt.Errorf("could not finish transfer even after %d attempts, lastErr: %w", maxReconnectAttempts, err)
+		}
+		duration := b.Duration()
+		bt := time.NewTimer(duration)
+		fmt.Println(duration)
```

Comment:
Suggest adding log lines for each attempt (and remove the fmt.Println())

Reply:
We will do it as part of the logging work.